### PR TITLE
Add datastore proxy for separating context

### PR DESCRIPTION
**Pull request by jzelinskie**
This removes the footguns for datastores that need to separate context before performing read-only actions.

### CODE REVIEW

File: internal/datastore/crdb/tx_test.go [hunk 1]
```diff
@@ -117,7 +117,7 @@ func TestTxReset(t *testing.T) {
 			require := require.New(t)
 
 			ds := b.NewDatastore(t, func(engine, uri string) datastore.Datastore {
-				ds, err := NewCRDBDatastore(
+				ds, err := newCRDBDatastore(
```
Comment (vroldanbet) [inline internal/datastore/crdb/tx_test.go:120]: do we not want the tests with the proxy wrapper that we will be run in practice?

File: internal/datastore/proxy/caching.go [hunk 2]
```diff
@@ -80,7 +80,7 @@ func (r *nsCachingReader) ReadNamespace(
 		loadedRaw, err, _ = r.p.readNsGroup.Do(nsRevisionKey, func() (any, error) {
 			// sever the context so that another branch doesn't cancel the
 			// single-flighted namespace read
-			loaded, updatedRev, err := r.Reader.ReadNamespace(datastore.SeparateContextWithTracing(ctx), nsName)
+			loaded, updatedRev, err := r.Reader.ReadNamespace(SeparateContextWithTracing(ctx), nsName)
```
Comment (vroldanbet) [inline internal/datastore/proxy/caching.go:83]: Why are we leaving this one here?

File: internal/datastore/proxy/context.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+// This is useful for datastores that do not want to close connections when a
+// cancel or deadline occurs.
+func NewSeparatingContextDatastoreProxy(d datastore.Datastore) datastore.Datastore {
+	return &ctxProxy{d}
+}
+
+type ctxProxy struct{ delegate datastore.Datastore }
+
+func (p *ctxProxy) ReadWriteTx(ctx context.Context, f datastore.TxUserFunc) (datastore.Revision, error) {
+	return p.delegate.ReadWriteTx(ctx, f)
```
Comment (vroldanbet) [inline internal/datastore/proxy/context.go:42]: maybe clarify we explicitly not want to sever context in operations that may involve writes?

**Review by vroldanbet — APPROVED** [review]
This is a nice improvement, thanks a ton! Left just some non-blocking nits, LGTM!

File: internal/datastore/proxy/caching.go (hunk 2 above)
Comment (jzelinskie) [reply internal/datastore/proxy/caching.go:83]: We actually don't want to cancel this -- it explicitly has a test to assert this behavior.

File: internal/datastore/crdb/tx_test.go (hunk 1 above)
Comment (jzelinskie) [reply internal/datastore/crdb/tx_test.go:120]: The tests use private fields on the datastore so we need to do this. No tests rely on cancellation.

**Pull request merged**